Fix virt-manager on non-gnome systems

aske opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

aske committed:
spice-gtk: propagate gstreamer plugins

Note: there is no sound without gst-plugins-good.

abbradar commented:
@grahamcofborg eval
@grahamcofborg build virtmanager

worldofpeace commented on pkgs/development/libraries/spice-gtk/default.nix:
```diff
@@ -87,10 +87,13 @@ stdenv.mkDerivation rec {
   ];
 
+  propagatedBuildInputs = [
+    gst_all_1.gst-plugins-base
```
I don't see a reason why this needed to be propagated. Perchance explain?

aske replied:
When SpiceGtkClient is loaded via GI it also needs GStreamer and cannot find it because needed paths are not in `GI_TYPELIB_PATH`. I was able to see the cause only after patching `virtManager/details/viewer.py` to show all exceptions.

offlinehacker replied:
Isn't this only required if spice is enabled? I thin it ahould the same as here, at least applying that change worked for me: https://github.com/NixOS/nixpkgs/blob/7e71cd829290b5a5521f7feef8c0c55d3eb087f0/pkgs/applications/virtualization/virt-viewer/default.nix#L31

offlinehacker replied:
Ahh nevermind, I can see that you patched spice-gtk directly, so should we remove these additional dependencies also in `virt-viewer`?

aske replied:
Yeah, `gst-plugins-base` could be removed, though there's also `gst-plugins-good`.
I'll add a commit after checking if it is needed.

worldofpeace commented:
This appears to be slightly expanded from #64618

aske commented:
Oh, thanks, missed that one.
virt-manager now works fine for me after these changes, previously I saw the same errors as in #62647.

abbradar commented:
Can confirm that fixes the issue, thanks!

worldofpeace commented:
> When SpiceGtkClient is loaded via GI it also needs GStreamer and cannot find it because needed paths are not in GI_TYPELIB_PATH. I was able to see the cause only after patching virtManager/details/viewer.py to show all exceptions.

I think we're preferring, in this case, to then make the typelibs available within it by hardcoding them with
a patch to the source code.

aske commented:
I'm not sure if this is possible given GObject Introspection dynamic nature, any ideas how it should be done?

emmanuelrosa commented:
I use `virt-manager` with the `i3` window manager, on NixOS. I confirm that this pull requests fixes two issues. I can once again:

1. Use Spice (instead of VNC) to connect to VMs
2. Open VMs; to edit and/or run them.

Thank you for this fix :)

astsmtl commented on pkgs/applications/virtualization/virt-manager/default.nix:
```diff
@@ -25,7 +26,7 @@ python3Packages.buildPythonApplication rec {
   buildInputs = [
     libvirt-glib vte dconf gtk-vnc gnome3.adwaita-icon-theme avahi
-    gsettings-desktop-schemas libosinfo gtk3
+    gsettings-desktop-schemas libosinfo gtk3 gtksourceview4 librsvg
```
Do we really need `librsvg` here?

aske commented on pkgs/applications/virtualization/virt-manager/default.nix:
```diff
@@ -25,7 +26,7 @@ python3Packages.buildPythonApplication rec {
   buildInputs = [
     libvirt-glib vte dconf gtk-vnc gnome3.adwaita-icon-theme avahi
-    gsettings-desktop-schemas libosinfo gtk3
+    gsettings-desktop-schemas libosinfo gtk3 gtksourceview4 librsvg
```
Actually yes, it would make more sense to move `wrapGAppsHook` to `buildInputs` as it already propagates `librsvg`.
The problem is that virtmanager tries to update icon cache in the `setup.py`.
I'm not sure how to pass a flag to disable this behavior to `self.distribution` in the `setup.py`.

I can move `gtk3` to `nativeBuildInputs` to make it work, but first want to try doing this via the `setup.py` flag.

worldofpeace replied:
All you need to do here is
```
setupPyBuildFlags = [
  "--no-update-icon-cache"
];
```

We also should be able to drop the bit about compiling schemas below since that's done in `setup.py`.

aske replied:
Unfortunately that would not fork as it passes flags to `build_ext` while virtmanager's `setup.py` checks `self.distribution`.

I've managed to make it work via custom `buildPhase`:
```
  buildPhase = ''
    runHook preBuild
    cp ${setuppy} nix_run_setup
    ${python3Packages.python.pythonForBuild.interpreter} nix_run_setup --no-update-icon-cache build_ext bdist_wheel
    runHook postBuild
  '';
```

Maybe we need to add a new `setupPyDistFlags` and modify build phases to include
```
... nix_run_setup ${(lib.concatStringsSep " " setupPyBuildFlags)} ... build_ext ...
```

@FRidh What do you think about this? I can make a PR if it's an acceptable change. Would incur mass rebuild though.

worldofpeace replied:
Sounds good to me @aske, I noticed that shortly after commenting this.

aske committed:
virtmanager: add missing GI dependencies

aske committed:
virtmanager: fix librsvg propagation from wrapGAppsHook

Custom buildPhase should be removed once there is support
for setupPyDistFlags (not a final name).

aske committed:
virtmanager: remove redundant gsettings schema compilation

Schemas are now compiled in the setup.py.

aske committed:
virtviewer: remove gstreamer plugins already propagated by spice-gtk

aske committed:
virtmanager: remove redundant gtk3 dependency

aske commented:
I think this PR is ready for now.

I've added a custom `buildPhase` so there's no need for mass-rebuild right now (if/when the change for `setupPyDistFlags` is accepted we can remove the phase).

Also added  `gst-plugins-good` to the `spice-gtk`'s `propagatedBuildInputs` as I found out while testing `virt-viewer` (thanks to @offlinehacker for directing attention there) sound was not working in either virtmanager or virtviewer without "good" plugins.

Funnily enough while testing I forgot to add `gtk3` back to `buildInputs` but it seems everything works fine without it.

offlinehacker commented:
Looks good to me, thanks for all the effort @aske  :)

offlinehacker commented:
Ha, haven't thought https://github.com/NixOS/nixpkgs/pull/64682 is going to be merged so fast. @aske can you please make additional pull request without custom `buildPhase`?

aske commented:
Yeah, I'll make a PR, though we have some time before it's merged into master from staging.

offlinehacker commented:
Ok, yea i can see now it's for staging, fine then :)
